go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -719,82 +719,150 @@ func installDependenciesAndBuild() {
+func aboveSupportedRange(version string) bool {
+	return semver.Compare(semver.MajorMinor("v"+version), "v"+maxGoVersion) > 0
+}
+
 // Check if `version` is lower than `minGoVersion` or higher than `maxGoVersion`. Note that for
 // this comparison we ignore the patch part of the version, so 1.20.1 and 1.20 are considered
 // equal.
 func outsideSupportedRange(version string) bool {
-	short := semver.MajorMinor("v" + version)
-	return semver.Compare(short, "v"+minGoVersion) < 0 || semver.Compare(short, "v"+maxGoVersion) > 0
+	return belowSupportedRange(version) || aboveSupportedRange(version)
 }
 
-// Check if `v.goModVersion` or `v.goEnvVersion` are outside of the supported range. If so, emit
-// a diagnostic and return an empty version to indicate that we should not attempt to install a
-// different version of Go.
-func checkForUnsupportedVersions(v versionInfo) (msg, version string) {
-	if v.goModVersionFound && outsideSupportedRange(v.goModVersion) {
-		msg = "The version of Go found in the `go.mod` file (" + v.goModVersion +
-			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion +
-			"). Writing an environment file not specifying any version of Go."
-		version = ""
-		diagnostics.EmitUnsupportedVersionGoMod(msg)
-	} else if v.goEnvVersionFound && outsideSupportedRange(v.goEnvVersion) {
-		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
-			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion +
-			"). Writing an environment file not specifying any version of Go."
+// Assuming `v.goModVersionFound` is false, emit a diagnostic and return the version to install,
+// or the empty string if we should not attempt to install a version of Go.
+func getVersionWhenGoModVersionNotFound(v versionInfo) (msg, version string) {
+	if !v.goEnvVersionFound {
+		// We definitely need to install a version. We have no indication which version was
+		// intended to be used to build this project. Go versions are generally backwards
+		// compatible, so we install the maximum supported version.
+		msg = "No version of Go installed and no `go.mod` file found. Writing an environment " +
+			"file specifying the maximum supported version of Go (" + maxGoVersion + ")."
+		version = maxGoVersion
+		diagnostics.EmitNoGoModAndNoGoEnv(msg)
+	} else if outsideSupportedRange(v.goEnvVersion) {
+		// We definitely need to install a version. We have no indication which version was
+		// intended to be used to build this project. Go versions are generally backwards
+		// compatible, so we install the maximum supported version.
+		msg = "No `go.mod` file found. The version of Go installed in the environment (" +
+			v.goEnvVersion + ") is outside of the supported range (" + minGoVersion + "-" +
+			maxGoVersion + "). Writing an environment file specifying the maximum supported " +
+			"version of Go (" + maxGoVersion + ")."
+		version = maxGoVersion
+		diagnostics.EmitNoGoModAndGoEnvUnsupported(msg)
```

Comment:
I was thinking about whether to suggest this during my last review, but I am not convinced that this is the right approach. Consider a self-hosted runner / custom image where a particular version of Go is installed on purpose. The behaviour that is implemented here would override whatever version of Go might be deliberately installed on the system. Perhaps we could do this if we find ourselves in a GHA environment (i.e. `GITHUB_ACTIONS` is set) but not otherwise?

Reply:
We try to use the version of Go in the environment if we can. This branch is dealing with the case that the version of Go in the environment is not supported, e.g. Go 1.21 has been released and we haven't added support for it yet.